tests/tool_cloudmetrics_collect_metrics_test.php

```diff
@@ -0,0 +1,170 @@
+     *
+     * @param array $items
+     */
+    public function send_metrics(array $items) {
+        $names = [];
+        foreach ($items as $item) {
+            $names[] = $item->name;
+        }
+        $this->mock->receive($names);
+    }
+}
+
+/**
+ * Test for collect_metrics_task.
+ */
+class tool_cloudmetrics_collect_metrics_test extends \advanced_testcase {
+
+    /**
+     * Set up before each test
+     */
+    protected function setUp(): void {
+        parent::setUp();
+        $this->resetAfterTest();
+    }
+
+    /**
+     * Test the get_frequency_cutoff method
+     *
+     * @dataProvider cutoff_provider
+     * @param $input
+     * @param $expected
+     */
+    public function test_get_frequency_cutoff($input, $expected) {
+        $task = new collect_metrics_task();
+        $this->assertEquals($expected, $task->get_frequency_cutoff($input));
+    }
+
+    /**
+     * Provider function for test_get_frequency_cutoff.
+     *
+     * @return array[]
+     */
+    public function cutoff_provider() {
+        return [
+            [ 5, manager::FREQ_5MIN ],
+            [ 15, manager::FREQ_15MIN ],
+            [ 30, manager::FREQ_30MIN ],
+            [ 60, manager::FREQ_HOUR ],
+            [ 180, manager::FREQ_3HOUR ],
+            [ 720, manager::FREQ_12HOUR ],
+            [ 1440, manager::FREQ_DAY ],
+            [ 10080, manager::FREQ_WEEK ],
+            [ 16, manager::FREQ_MIN ],
+            [ 724, manager::FREQ_MIN ],
+            [ 2, manager::FREQ_MIN ],
+            [ 90, manager::FREQ_30MIN ],
+        ];
+    }
+
+    /**
+     * Test the execute method.
+     *
+     * @dataProvider execute_provider
+     * @param int $time
+     * @param array $expected
+     * @throws \Exception
+     */
+    public function test_execute($timestr, $freqs, $expected) {
+        set_config('onlineusers_frequency', $freqs[0], 'tool_cloudmetrics');
+        set_config('activeusers_frequency', $freqs[1], 'tool_cloudmetrics');
+        set_config('newusers_frequency', $freqs[2], 'tool_cloudmetrics');
+        set_config('activeusers_enabled', 1, 'tool_cloudmetrics');
+        set_config('onlineusers_enabled', 1, 'tool_cloudmetrics');
+        set_config('newusers_enabled', 1, 'tool_cloudmetrics');
+
+        $tz = \core_date::get_server_timezone_object();
+        $time = (new \DateTime($timestr, $tz))->getTimestamp();
+
+        $mock = $this->createMock(mock_receiver::class);
+        $mock->expects($this->once())
+            ->method('receive')
+            ->with($expected);
+        $task = new helper_collect_metrics_task($mock);
+        $task->set_time($time);
+        $task->execute();
+    }
+
+    /**
+     * Provider function for test_execute.
+     *
+     * @return array[] Each element contains a time string, a list of frequency settings,
+     *                 and a list of metrics that should be measured.
+     */
+    public function execute_provider(): array {
```

Comment:
Kevin recently did some tests like this, using a generator instead, and got really wide coverage. Could we look at doing something like that?

https://github.com/catalyst/moodle-tool_mfa/blob/master/factor/token/tests/factor_test.php#L260

Reply:
I don't think that would work here. How would you determine what metrics are expected to be included?

Reply:
Was just a thought, there are intricacies here that go over my head atm

Reply:
the generator method would just need to be smart, but its probably not worth